src/lib.rs

```diff
@@ -1,70 +1,98 @@
-                ("crate", "".to_string())
-            } else if let Some(repo) = repo {
-                detect_repo(&repo)?
-            } else {
-                ("crate", "".to_string())
-            }
-        }
-        "crate" => ("crate", "".to_string()),
-        _ => {
-            if repo.is_none() {
-                bail!("Could not find repository path in crates.io.");
-            }
-            (method_name, repo.unwrap())
+#[macro_use]
+extern crate log;
+
+static APP_USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"));
+
+/// https://api.bitbucket.org/2.0/repositories
+pub const DEFAULT_BITBUCKET_URL: &'static str = "https://api.bitbucket.org/2.0/repositories";
+/// https://github.com
+pub const DEFAULT_GITHUB_URL: &'static str = "https://github.com";
+/// https://gitlab.com
+pub const DEFAULT_GITLAB_URL: &'static str = "https://gitlab.com";
+/// https://crates.io
+pub const DEFAULT_REGISTRY_URL: &'static str = "https://crates.io";
+
+/// An enum representation of supported cloning methods.
+#[derive(Debug, Clone)]
+pub enum CloneMethodKind {
+    Crate,
+    Git,
+    Mercurial,
+    Pijul,
+    Fossil,
+    Auto,
+}
+
+impl CloneMethodKind {
+    /// Returns the underlying command line command for the method.
+    pub fn command(&self) -> &str {
+        match *self {
+            CloneMethodKind::Crate => "crate",
+            CloneMethodKind::Git => "git",
+            CloneMethodKind::Mercurial => "hg",
+            CloneMethodKind::Pijul => "pijul",
+            CloneMethodKind::Fossil => "fossil",
+            CloneMethodKind::Auto => "auto",
         }
-    };
-    match method {
-        "crate" => {
-            if !extra.is_empty() {
-                bail!("Got extra arguments, crate downloads take no extra arguments.");
-            }
-            clone_crate(name, version_req, &pkg_info)?;
+    }
+
+    /// Creates a `CloneMethodKind` from a method name. If no name matches then None is returned.
+    /// Current options are `crate`, `git`, `hg`, `mercurial`, `pijul`, `fossil`, and `auto`
+    pub fn from(method_name: &str) -> Option<CloneMethodKind> {
+        match method_name {
+            "crate" => Some(CloneMethodKind::Crate),
+            "git" => Some(CloneMethodKind::Git),
+            "hg" => Some(CloneMethodKind::Mercurial),
+            "mercurial" => Some(CloneMethodKind::Mercurial),
+            "pijul" => Some(CloneMethodKind::Pijul),
+            "fossil" => Some(CloneMethodKind::Fossil),
+            "auto" => Some(CloneMethodKind::Auto),
+            _ => None,
         }
-        "git" | "hg" | "pijul" | "fossil" => {
-            if let Some(version_req) = version_req {
-                bail!(
-                    "Specifying a version `{}` only works with the `crate` method.",
-                    version_req
-                );
-            }
-            run_clone(method, &repo, extra)?;
+    }
+}
+
+impl std::fmt::Display for CloneMethodKind {
```

Comment:
This display impl doesn't seem to be really necessary.

Reply:
Updated